src/kernel/boot.c

```diff
@@ -599,6 +628,29 @@ BOOT_CODE static bool_t provide_untyped_cap(
 {
     bool_t ret;
     cap_t ut_cap;
+
+    /* Since we are in boot code, we can do extensive error checking and
+       return failure if anything unexpected happens. */
+
+    /* Bounds check for size parameter */
+    if (size_bits > seL4_MaxUntypedBits || size_bits < seL4_MinUntypedBits) {
+        printf("Kernel init: Invalid untyped size %lu", size_bits);
+        return false;
+    }
+
+    /* All cap ptrs must be aligned to object size */
+    if (!IS_ALIGNED(pptr, size_bits)) {
+        printf("Kernel init: Unaligned untyped paddr %p (alignedment %lu)", (void *)pptr, size_bits);
+        return false;
+    }
+
+    /* All cap ptrs apart from device untypeds must be in the kernel window */
+    if (!device_memory && !pptr_in_kernel_window(pptr)) {
+        printf("Kernel init: Non-device untyped paddr %p outside kernel window",
+               (void *)pptr);
+        return false;
+    }
```

Comment:
Does this also need to check that the end of the region isn't outside of the kernel window? I don't think there's a constraint that the kernel window is power-of-2 sized.

Reply:
Yes, good point, will add that check.

Reply:
Done.